pkg/clusterapi/etcd.go

```diff
@@ -0,0 +1,86 @@
+package clusterapi
+
+import (
+	etcdbootstrapv1 "github.com/aws/etcdadm-bootstrap-provider/api/v1beta1"
+	etcdv1 "github.com/aws/etcdadm-controller/api/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+)
+
+// SetEtcdConfigInEtcdCluster sets up the etcd config in EtcdadmCluster.
+func SetEtcdConfigInEtcdCluster(etcd *etcdv1.EtcdadmCluster, version string) {
+	etcd.Spec.EtcdadmConfigSpec.CloudInitConfig = &etcdbootstrapv1.CloudInitConfig{
+		Version:    version,
+		InstallDir: "/usr/bin",
+	}
+	etcd.Spec.EtcdadmConfigSpec.PreEtcdadmCommands = []string{
+		"hostname \"{{`{{ ds.meta_data.hostname }}`}}",
+		"echo \"::1         ipv6-localhost ipv6-loopback\" >/etc/hosts",
+		"echo \"127.0.0.1   localhost\" >>/etc/hosts",
+		"echo \"127.0.0.1   {{`{{ ds.meta_data.hostname }}`}}\" >>/etc/hosts",
+		"echo \"{{`{{ ds.meta_data.hostname }}`}}\" >/etc/hostname",
+	}
+}
+
+// SetEtcdConfigInCluster sets up the etcd config in CAPI Cluster.
+func SetEtcdConfigInCluster(cluster *clusterv1.Cluster, clusterSpec *cluster.Spec) {
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration == nil {
+		return
+	}
+
+	cluster.Spec.ManagedExternalEtcdRef = &v1.ObjectReference{
+		APIVersion: etcdClusterAPIVersion,
+		Kind:       etcdadmClusterKind,
+		Name:       EtcdClusterName(clusterSpec),
+		Namespace:  constants.EksaSystemNamespace,
+	}
+}
+
+// SetUnstackedEtcdConfigInKubeadmControlPlaneForBottlerocket sets up unstacked etcd configuration in kubeadmControlPlane for bottlerocket.
+func SetUnstackedEtcdConfigInKubeadmControlPlaneForBottlerocket(kcp *controlplanev1.KubeadmControlPlane, externalEtcdConfig *v1alpha1.ExternalEtcdConfiguration) {
+	if externalEtcdConfig == nil {
```

Comment:
why not perform this check in the caller and that way you don't need to pass `externalEtcdConfig`?

Reply:
why not passing this in the method so the caller does not need to have the if condition? we use same pattern for all methods in apibuilder and imo it is clear: updating etcd config in kcp based on externalEtcdConfig arg.